compiler/src/main/kotlin/io/github/aplcornell/viaduct/backends/cleartext/CleartextCircuitCodeGenerator.kt

```diff
@@ -112,6 +121,128 @@ class CleartextCircuitCodeGenerator(context: CodeGeneratorContext) : AbstractCod
         }
     }
 
+    private fun createCommitment(
+        source: Protocol,
+        target: Protocol,
+        argument: Argument,
+        builder: CodeBlock.Builder,
+    ): CodeBlock {
+        require(context.host in source.hosts + target.hosts)
+        if (source !is Local) {
+            throw UnsupportedCommunicationException(source, target, argument.sourceLocation)
+        }
+        require(source.hosts.size == 1 && source.host in source.hosts)
+        require(target is CommitmentProtocol)
+        if (target.cleartextHost != source.host || target.cleartextHost in target.hashHosts) {
+            throw UnsupportedCommunicationException(source, target, argument.sourceLocation)
+        }
+
+        val argType = kotlinType(argument.type.shape, typeTranslator(argument.type.elementType.value))
+        val sendingHost = target.cleartextHost
+        val receivingHosts = target.hashHosts
+        return when (context.host) {
+            sendingHost -> {
+                val tempName1 = context.newTemporary("CommitTemp")
+                val tempName2 = context.newTemporary("CommitTemp")
+                builder.addStatement(
+                    "val %N = %T(%L)",
+                    tempName1,
+                    (Committed::class).asTypeName().parameterizedBy(argType),
+                    argument.value,
+                )
+                builder.addStatement(
+                    "val %N = %N.%M()",
+                    tempName2,
+                    tempName1,
+                    MemberName(Committed.Companion::class.asClassName(), "commitment"),
+                )
+                receivingHosts.forEach {
+                    builder.addStatement("%L", context.send(CodeBlock.of("%N", tempName2), it))
+                }
+                CodeBlock.of("%N", tempName1)
+            }
+
+            in receivingHosts -> {
+                val tempName3 = context.newTemporary("CommitTemp")
+                builder.addStatement(
+                    "val %N = %L",
+                    tempName3,
+                    context.receive((Commitment::class).asTypeName().parameterizedBy(argType), source.host),
+                )
+                CodeBlock.of("%N", tempName3)
+            }
+
+            else -> throw IllegalStateException()
+        }
+    }
+
+    private fun openCommitment(
+        source: Protocol,
+        target: Protocol,
+        argument: Argument,
+        builder: CodeBlock.Builder,
+    ): CodeBlock {
+        require(source is CommitmentProtocol)
+        if (target !is Cleartext) {
+            throw UnsupportedCommunicationException(source, target, argument.sourceLocation)
+        }
+        require(context.host in source.hosts + target.hosts)
+        if (source.hashHosts != target.hosts || source.cleartextHost in source.hashHosts) {
```

Comment:
You don't need to check `source.cleartextHost in source.hashHosts`, that's ensured by `Commitment`.